src/sentry/api/endpoints/user_authenticator_enroll.py

```diff
@@ -0,0 +1,202 @@
+            serialized_field['defaultValue'] = serializer.base_fields[field].get_default_value()
+            meta.append(serialized_field)
+
+    return meta
+
+
+class UserAuthenticatorEnrollEndpoint(UserEndpoint):
+    permission_classes = (IsAuthenticated, UserPermission, )
+
+    # @sudo_required
+    def get(self, request, user, interface_id):
+        """
+        Get Authenticator Interface
+        ```````````````````````````
+
+        Retrieves authenticator interface details for user depending on user enrollment status
+
+        :pparam string user_id: user id or "me" for current user
+        :pparam string interface_id: interface id
+
+        :auth: required
+        """
+
+        interface = Authenticator.objects.get_interface(user, interface_id)
+
+        # Not all interfaces allow multi enrollment
+        if interface.is_enrolled and not interface.allow_multi_enrollment:
+            return Response(status=status.HTTP_400_BAD_REQUEST)
+
+        # User is not enrolled in auth interface:
+        # - display configuration form
+        response = serialize(interface)
+        response['form'] = get_serializer_field_metadata(
+            serializer_map[interface_id]
+        )
+
+        # U2fInterface has no 'secret' attribute
+        try:
+            response['secret'] = interface.secret
+        except AttributeError:
+            pass
+
+        if interface_id == 'totp':
+            response['qrcode'] = interface.get_provision_qrcode(user.email)
+
+        if interface_id == 'u2f':
+            response['challenge'] = interface.start_enrollment()
+
+        return Response(response)
+
+    @sudo_required
+    def post(self, request, user, interface_id):
+        """
+        Enroll in authenticator interface
+        `````````````````````````````````
+
+        :pparam string user_id: user id or "me" for current user
+        :pparam string interface_id: interface id
+
+        :auth: required
+        """
+
+        # start activation
+        serializer_cls = serializer_map.get(interface_id, None)
+
+        if serializer_cls is None:
+            return Response(status=status.HTTP_404_NOT_FOUND)
+
+        serializer = serializer_cls(data=request.DATA)
+
+        if not serializer.is_valid():
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+        interface = Authenticator.objects.get_interface(user, interface_id)
+
+        # Not all interfaces allow multi enrollment
+        #
+        # This is probably un-needed because we catch
+        # `Authenticator.AlreadyEnrolled` when attempting to enroll
+        if interface.is_enrolled and not interface.allow_multi_enrollment:
+            return Response(status=status.HTTP_400_BAD_REQUEST)
+
+        try:
+            interface.secret = request.DATA['secret']
+        except KeyError:
+            pass
+
+        # Need to update interface with phone number before validating OTP
+        if 'phone' in request.DATA:
```

Comment:
Is this odd that API endpoint behaves differently based on request params?

Reply:
nah i think it's ok